### ISSUE TITLE
Best way to receive the return value when the type hint is Union[Awaitable[int], int]?

### ISSUE THREAD

**Issue by hyperstarkkk** [redis/redis-py#2399]
I noticed the return value type hint of redis-py functions are always like: Union[Awaitable[int], int]: (e.g. [redis.commands.core.py:L4824](https://github.com/redis/redis-py/blob/master/redis/commands/core.py#L4824))

```
def hlen(self, name: str) -> Union[Awaitable[int], int]:
    """
    Return the number of elements in hash ``name``

    For more information see https://redis.io/commands/hlen
    """
    return self.execute_command("HLEN", name)
```
When I receive the value as normal
```
dict_size = redis.hlen("dict_key")
assert dict_size < 10
```
Pylance will remind me
```
Operator "<=" not supported for types "Awaitable[int] | int" and "Literal[10]"
  Operator "<=" not supported for types "Awaitable[int]" and "Literal[10]"
```
How to let Pylance know I am not calling the function in asynchronous way, so I can mitigate this type checking error?

**Comment by akx** [comment]
That smells like a typing bug in the library. `Awaitable`s should really only be returned when using the async client...

**Comment by balazser** [comment]
I'm also facing this issue. I there any update?

```python
    def llen(self, name: str) -> Union[Awaitable[int], int]:
        """
        Return the length of the list ``name``

        For more information see https://redis.io/commands/llen
        """
```
Many of the methods with `Union[Awaitable[int], int]` require a type guard check, or what is the best way to get around it? :thinking:

**Comment by Kyle-sandeman-mrdfood** [comment]
Similarly for almost ALL functions in commands/core.py, they're annotated to return `ResponseT` (or similar) which is `Union[Awaitable, Any]` which is picked up by PyCharm as just `Awaitable`. So passing e.g. `r.get(key)`'s result to anything that expects bytes/str results in a warning

**Comment by divad** [comment]
I'm hitting this, but in the reverse; mypy complains when using the asyncio version that it might return int, even when I'm importing from redis.asyncio and calling `await` against the function. Any update?

**Comment by armarik** [comment]
I'm also facing the same issue.

**Comment by honzaflash** [comment]
Seems like this should be mentioned here. One of the linked issues above contains a reply with a solution/workaround:
https://github.com/redis/redis-py/issues/3091#issuecomment-1903419256

The suggested solution is to install the type stub package: https://pypi.org/project/types-redis/

Worked for me :+1: 

Afaiu, this is the old way of getting typing hints for redis but the package now has its own typing hints that are incomplete. Because of that pylance won't tell you to go install available type stubs but it will complain because the included types return types are wrong.

**Comment by umstek** [comment]
> Seems like this should be mentioned here. One of the linked issues above contains a reply with a solution/workaround: [#3091 (comment)](https://github.com/redis/redis-py/issues/3091#issuecomment-1903419256)
> 
> The suggested solution is to install the type stub package: https://pypi.org/project/types-redis/
> 
> Worked for me 👍
> 
> Afaiu, this is the old way of getting typing hints for redis but the package now has its own typing hints that are incomplete. Because of that pylance won't tell you to go install available type stubs but it will complain because the included types return types are wrong.

But it says

> The redis package includes type annotations or type stubs since version 5.0.0. Please uninstall the types-redis package if you use this or a newer version.

although the issue still exists on versions > 5

**Comment by aa0308qq** [comment]
For the latest version of `redis-py`, the approach I use is:
```
from typing import Any, Awaitable
from redis.commands.core import ResponseT

class RedisResult:
    @classmethod
    def extract(cls, value: ResponseT) -> Any:
        if isinstance(value, Awaitable):
            raise TypeError
        return value

    @classmethod
    async def extract_async(cls, value: ResponseT) -> Any:
        if not isinstance(value, Awaitable):
            raise TypeError
        return await value
``` 
For `sync`:
```
from typing import Union
from redis import Redis

sync_redis = Redis(host=host, port=port)
task:Union[str,None] = RedisResult.extract(sync_redis.hget("task","uuid"))
```
For `async`:
```
from typing import Union
from redis.asyncio import Redis as AsyncRedis

async_redis = AsyncRedis(host='localhost', port=port)
task:Union[str,None] =await RedisResult.extract_async(async_redis.hget("task","uuid"))
```

**Comment by ykun91** [comment]
> irritating

Exactly. I don't think Redis, Inc maintain this redis-py well, the code is bad typed in general and lots of issues seems to be neglected.

May better for this project to delete the `py.typed` and accept the community maded typeshed, instead of releasing the half-baked types and wasting users time.

**Comment by ykun91** [comment]
> > Seems like this should be mentioned here. One of the linked issues above contains a reply with a solution/workaround: [#3091 (comment)](https://github.com/redis/redis-py/issues/3091#issuecomment-1903419256)
> > The suggested solution is to install the type stub package: https://pypi.org/project/types-redis/
> > Worked for me 👍
> > Afaiu, this is the old way of getting typing hints for redis but the package now has its own typing hints that are incomplete. Because of that pylance won't tell you to go install available type stubs but it will complain because the included types return types are wrong.
> 
> Thanks this has fixed it :)
> 
> It's always a bit irritating when something as big as redis has typing support, but it's incomplete in a way that it fails for the most basic stuff. So I'll leave this here as a push :D

Exactly. I don't think Redis, Inc maintain this redis-py well, the code is bad typed in general and lots of issues seems to be neglected.

May better for this project to delete the `py.typed` and accept the community maded typeshed, instead of releasing the half-baked types and wasting users time.

**Comment by zouri** [comment]
``` python
res = rdb.sadd(key, *values)
assert isinstance(res, Awaitable)
return await res
```
Check the return value for its type to circumvent static checks. Of course, if you're not confident, you can also replace it with an 'if' statement.

**Comment by imnotjames** [comment]
The type stubs at `types-redis` have been [removed from typeshed](https://github.com/python/typeshed/pull/13157) and will no longer see updates.  Thankfully, they aren't going to yank the packages yet.

Perhaps the types should be dropped entirely from this project, as they are basically not usable & it seems there's no interest in improving them from Redis Labs, Inc?  There's a few open pull requests that have been open for months with no traction.

> Exactly. I don't think Redis, Inc maintain this redis-py well, the code is bad typed in general and lots of issues seems to be neglected.
>
> May better for this project to delete the py.typed and accept the community maded typeshed, instead of releasing the half-baked types and wasting users time.

Agreed, but typeshed removed it because they are concerned about Redis Labs, Inc taking legal action against them -- so I don't think there will be movement to reintroduce them.

**Comment by anchit-sadana** [comment]
Has anyone at Redis said anything about improving type hints for this package beyond the threatening legal action against typeshed?

**Comment by imnotjames** [comment]
> Has anyone at Redis said anything about improving type hints for this package beyond the threatening legal action against typeshed?

To clarify: No legal action or threats were sent out to typeshed to my knowledge.  It was mostly concerns because there is a belief out in the wild that redis is ["tightening their use of the trademark"](https://github.com/python/typeshed/issues/10592#issuecomment-2506735812).

The typeshed types weren't great, either - they were stuck on redis-py 4.  According to the [relevant typeshed issue](https://github.com/python/typeshed/issues/10592#issuecomment-2506855987), they were hoping for support or cooperation but hadn't seen any.  And many [issues](https://github.com/redis/redis-py/issues/2730) [relating](https://github.com/redis/redis-py/issues/3195) [to types](https://github.com/redis/redis-py/issues/2933) are not really seeing feedback from anyone that is a maintainer that I can see.

However, without feedback from a redis engineer / maintainer we don't have much as far as info on what their roadmap is.  Some bug fixes & improvements for Redis enterprise features maybe?

**Comment by JohnRichard4096** [comment]
Hmm, this issue still hasn’t been resolved even in **2025**(after three years,).
Perhaps the best workaround is to rely on `typing.cast`, `assert`, or `# type: ignore`—but overusing these can clutter the code with excessive assertions, which is far from ideal.

And this is just one of many such problems...

**Comment by provinzkraut** [comment]
Updating this for the year 2026 so it's not believed to be stale.....

@petyaslavova, @vladvildanov, just pinging some active maintainers here in the hope to get *any* reaction on this issue that's been open 4 years now. Even a notice like "we're not planning to fix the typing" would be helpful to the community.

**Comment by petyaslavova** [comment]
Fixed with PRs #3991 and #4005

**Comment by Dude29** [comment]
@petyaslavova thank you!
